### ISSUE TITLE
Toolshelf -  add the text menus as panels

### ISSUE THREAD

**Issue by ReinerBforartists** [Bforartists/Bforartists#2402]
We had in Bforartists 1 already everything in the toolshelf that made sense. We could grab this panels, and port it to Bforartists 2.

It is of course a bit of work. We need to remove the one or another thing. We need to adjust the icon size, and so on. But we don't start at zero that way.

And once this task is done we can have a look where we want to go next.

![bfa1tabs](https://user-images.githubusercontent.com/13919702/117531166-7f1c5e00-afe1-11eb-8c44-fee67eac0273.jpg)

**Comment by ReinerBforartists** [comment]
I suggest to do this task tab by tab and panel by panel. A commit for each change ...

**Comment by ReinerBforartists** [comment]
Reminder, new panels goes now into space_toolsystem_toolbar_tabs.py

**Comment by ReinerBforartists** [comment]
Postponing. This task needs to be tackled after we have solved the main problems. Like transparent background etc. . Else we might do the double work.

**Comment by ReinerBforartists** [comment]
Let's add a few panels here. I think we should skip the create section though. We have the add menu content already in the toolbar.

**Comment by ReinerBforartists** [comment]
It's a bit more complicated than thought. Not done with copying over the panels. More recreate the panels. But helps.

**Comment by ReinerBforartists** [comment]
![transform](https://user-images.githubusercontent.com/13919702/118301404-4a2d7100-b4e3-11eb-8773-cc55c7940b1c.png)

**Comment by ReinerBforartists** [comment]
Now i know why it took so long with Bforartists 1. The modes are killing me XD

**Comment by ReinerBforartists** [comment]
Reminder, don't forget to link the set dimensions entry into the panel too. It's an addon ...

**Comment by Draise14** [comment]
> Now i remember why it took so long with Bforartists 1. The modes are killing me XD

Is there a way to split this up to modes maybe? Maybe as an easier log and chop the size to bite sized tasks?

**Comment by ReinerBforartists** [comment]
You would just increase the confusion by adding a panel for every mode. And the work is the same :)

**Comment by ReinerBforartists** [comment]
Well, i do it in small chunks actually. But this is one of the tasks that is better done by one person. To avoid conflicts in the implementation :)

**Comment by ReinerBforartists** [comment]
Renamed the classes to make the names more consistent, and to avoid conflicts with existing blender panels.

The name convention is now Prefix like in Blender. Then the tab plus "tab" at the end. Then panelname. That way we should not have conflicts with Blender panels. Example:

VIEW3D_PT_objecttab_snap

**Comment by ReinerBforartists** [comment]
The set dimensions addon appends. We don't have any influence at the button shape here. But i think we can live with this button shape.

![setdimensions](https://user-images.githubusercontent.com/13919702/118627596-295b6900-b7cc-11eb-87d1-9696b39f21a7.jpg)

**Comment by ReinerBforartists** [comment]
Phew, decision time. Now it's getting hellish complicated. We can go through the menus mode by mode and type by type. And use  bl_context = "XXXXXmode" to show the panel. This would be the cleanest method. And we can get away with this in most of the panels.

But this method can produce lots of panels in other cases. Alone the transform panel has minimum three modes with corresponding object types. We would end in 20 equal panels that way just for the transform panel.

Time for some experiments.

**Comment by ReinerBforartists** [comment]
Let's make a panel version for every mode by bl_context. It's simply cleaner than to go crazy with the def poll.

**Comment by ReinerBforartists** [comment]
It remains difficult. There is no bl_context editmode. There is a bl_context edit_mesh, edit_gpencil ....

I told you the modes are killing me here ^^

**Comment by ReinerBforartists** [comment]
https://docs.blender.org/api/current/bpy.context.html

bpy.context.mode
Type
enum in [‘EDIT_MESH’, ‘EDIT_CURVE’, ‘EDIT_SURFACE’, ‘EDIT_TEXT’, ‘EDIT_ARMATURE’, ‘EDIT_METABALL’, ‘EDIT_LATTICE’, ‘POSE’, ‘SCULPT’, ‘PAINT_WEIGHT’, ‘PAINT_VERTEX’, ‘PAINT_TEXTURE’, ‘PARTICLE’, ‘OBJECT’, ‘PAINT_GPENCIL’, ‘EDIT_GPENCIL’, ‘SCULPT_GPENCIL’, ‘WEIGHT_GPENCIL’, ‘VERTEX_GPENCIL’], default ‘EDIT_MESH’, (readonly)

Where the heck is object mode here? ^^

**Comment by ReinerBforartists** [comment]
For the set origin panel bl_context = "objectmode" made most sense ...

**Comment by ReinerBforartists** [comment]
Mirror. Wasn't there the problem that the local version does work global or such? I will have a look to make the local version in a sub panel, and closed ...
![mirror](https://user-images.githubusercontent.com/13919702/118938333-d100a480-b94e-11eb-87d5-14ec781a799e.jpg)

**Comment by ReinerBforartists** [comment]
Mirrorpanel. Local is a subpanel.

![mirror](https://user-images.githubusercontent.com/13919702/118950326-60f81b80-b95a-11eb-857a-19f5ac1be0f1.jpg)

**Comment by ReinerBforartists** [comment]
Clear panel

![clear](https://user-images.githubusercontent.com/13919702/118955276-d534be00-b95e-11eb-8048-53152a3c6b8d.jpg)

**Comment by iyadahmed** [comment]
The new panels are beautiful, and I just realized they might be touch friendly too!

**Comment by ReinerBforartists** [comment]
They should be :)

**Comment by ReinerBforartists** [comment]
Committed to the wrong issue. But don't want to revert the whole thing again. Apply panel ...

https://github.com/Bforartists/Bforartists/commit/60744685bdd2548336d7d2991a330bbd144a7c28

![apply](https://user-images.githubusercontent.com/13919702/119114443-75a1e580-ba26-11eb-9a70-5c404d905aea.jpg)

**Comment by ReinerBforartists** [comment]
Okay, i think we are through with the object menu related content. Let's go through the menus in the modes ...

**Comment by ReinerBforartists** [comment]
We need new icons for merge, separate and split. They all share the same icons each  at the moment.

![newicons](https://user-images.githubusercontent.com/13919702/119220242-e0682500-bae9-11eb-8d8b-a2b980ee92d3.jpg)

**Comment by ReinerBforartists** [comment]
Merge, Split and Separate in a panel called Parts.

The icon problem is separated into this issue: Edit Mode - Mesh menu - create separated icons for merge, split and separate #2446

![parts](https://user-images.githubusercontent.com/13919702/119221524-1f997480-baf0-11eb-8c82-9e44a61854ee.jpg)

**Comment by ReinerBforartists** [comment]
Tools panel in edit mode

![tools](https://user-images.githubusercontent.com/13919702/119222580-6342ad00-baf5-11eb-9f37-f679365300c7.jpg)

**Comment by ReinerBforartists** [comment]
Hm. I think i made a mistake to put merge separate and split into one panel. To have each in its own panel allows better fine control. And then you can pin the merge panel for example ...

I will separate these panels.

**Comment by ReinerBforartists** [comment]
And i made wrong labels XD

![wrongseparatelabels](https://user-images.githubusercontent.com/13919702/119462782-6631dd80-bd41-11eb-9969-92e274a0594e.jpg)

**Comment by ReinerBforartists** [comment]
Not sure if we should put normals into the sidebar. Too much options for a tool that you don't really use very regularly.

What we could put into an own panel is the most common operations. Recalc outside, inside and flip.

![fitsbetterintext](https://user-images.githubusercontent.com/13919702/119475159-2ffa5b00-bd4d-11eb-8933-3b4f3755e1cf.jpg)

**Comment by ReinerBforartists** [comment]
Hm. maybe we can add the exotic methods as sub panels.

Normals panel has landed.

![normalspanel](https://user-images.githubusercontent.com/13919702/119478254-f2e39800-bd4f-11eb-9b97-7afc1a1e0b5e.jpg)

**Comment by iyadahmed** [comment]
> Not sure if we should put all normals content into the sidebar. Too much options for a toolset that you don't really use very regularly. Especially the exotic methods.
> 
> 
> 
> What we could put into an own panel is the most common operations. Recalc outside, inside and flip.
> 
> 
> 
> ![fitsbetterintext](https://user-images.githubusercontent.com/13919702/119475159-2ffa5b00-bd4d-11eb-8933-3b4f3755e1cf.jpg)
> 
> 

Agree only common operators in panel, otherwise it would be too much duplication

**Comment by ReinerBforartists** [comment]
Object mode, shading tab

![shading](https://user-images.githubusercontent.com/13919702/119532752-753a7f00-bd85-11eb-96e6-cea8f7b83049.jpg)

**Comment by ReinerBforartists** [comment]
Shading in edit mode with a mesh object. I guess we should create four more icons here

![shading](https://user-images.githubusercontent.com/13919702/119609487-efa3e700-bdf7-11eb-9be6-b6ad2c97d904.jpg)

**Comment by ReinerBforartists** [comment]
New icons added. Let's make a panel ...

![shadingfixed](https://user-images.githubusercontent.com/13919702/119616047-6fce4a80-be00-11eb-8afd-995f245bc0e3.jpg)

**Comment by ReinerBforartists** [comment]
Shading icon for mesh in edit mode

![shading](https://user-images.githubusercontent.com/13919702/119619634-7ced3880-be04-11eb-9158-5107465d9166.jpg)

**Comment by ReinerBforartists** [comment]
Dissolve panel

![dissolve](https://user-images.githubusercontent.com/13919702/119622468-77ddb880-be07-11eb-81af-29c582e94ece.jpg)

**Comment by ReinerBforartists** [comment]
Vertex tab, vertex panel

![vertextab](https://user-images.githubusercontent.com/13919702/119624914-eb80c500-be09-11eb-93fb-c40f3721659f.jpg)

**Comment by ReinerBforartists** [comment]
wtf?

![wtf](https://user-images.githubusercontent.com/13919702/119629043-f2a9d200-be0d-11eb-9672-ffe6be66e137.jpg)

**Comment by ReinerBforartists** [comment]
Edge panel. No idea what this hiccup was. I made it now more compact. And hiccup is gone.

Needs three new icons. At work ...

![wtf](https://user-images.githubusercontent.com/13919702/119630553-584a8e00-be0f-11eb-90a8-3aa82ba86a8a.jpg)

**Comment by ReinerBforartists** [comment]
Added new icons

![icons](https://user-images.githubusercontent.com/13919702/119638138-a911b500-be16-11eb-8ee7-f87ae3201663.jpg)

**Comment by ReinerBforartists** [comment]
Face panel ...

![faces](https://user-images.githubusercontent.com/13919702/119693838-f3158d80-be4c-11eb-9470-0bc5135b5ea9.jpg)

**Comment by ReinerBforartists** [comment]
This should simply not happen. But does -.-

![shouldnthappen](https://user-images.githubusercontent.com/13919702/119695597-a29f2f80-be4e-11eb-9df3-151e9ecf4242.jpg)

**Comment by ReinerBforartists** [comment]
UV panel

![uv](https://user-images.githubusercontent.com/13919702/119696951-024a0a80-be50-11eb-8f55-003c437c2093.jpg)

**Comment by ReinerBforartists** [comment]
Project from view with bounds needs own icon.

**Comment by Draise14** [comment]
I'm really really loving these updates.

Just a request, can we also expose the "Mesh>Cleanup"  panel tool group? Those are very useful and I sometimes have that in a quick menu to compensate using them so often. Specially preparing meshes for game assets. Can be collapsed by default.

**Comment by ReinerBforartists** [comment]
Hm, cleanup is one of these toolsets that i would love to keep in the mesh menu. But okay.

By default, when we have more than one panel in the tab then just one panel is open.

**Comment by iyadahmed** [comment]
> Added new icons
> 
> 
> 
> ![icons](https://user-images.githubusercontent.com/13919702/119638138-a911b500-be16-11eb-8ee7-f87ae3201663.jpg)
> 
> 
I think with separators it will be better

**Comment by ReinerBforartists** [comment]
Okay, will have a look if i am able to.

**Comment by iyadahmed** [comment]
> Okay, will have a look if i am able to. But this will make the whole list much longer. And the code more complicated.

Oh ok

**Comment by ReinerBforartists** [comment]
Hmmhmmm. Technically of course possible. The problem is the three row layout then. With separators we would produce lots of empty space here. And this is no improvement really. There would not be a benefit to dragging it from two rows to three rows anymore. 

One row and text have the separators. Here it makes imho sense. But with two and three rows i would love to have it as compact as possible.

Long story short, i would like to leave it the current way for now. Maybe we can discuss this at a later point, and do refinement when needed.

![longlist](https://user-images.githubusercontent.com/13919702/119794274-ac20aa00-bed7-11eb-9e9d-4a6e66a23db4.jpg)

![longlist2](https://user-images.githubusercontent.com/13919702/119794289-af1b9a80-bed7-11eb-8a3b-6121176b3c4d.jpg)

**Comment by ReinerBforartists** [comment]
Heh, i made in fact the mistake to separate the icons in the tools panel. A nice example. Imagine how long the list will become when we separate the icons by line in all the other panels too :)

Tools panel needs a fix.

![toolsproblem](https://user-images.githubusercontent.com/13919702/119797451-8943c500-beda-11eb-8079-5045e54dd793.jpg)

**Comment by ReinerBforartists** [comment]
Transform and snap should also be compact.

![transformnsnap](https://user-images.githubusercontent.com/13919702/119798004-0cfdb180-bedb-11eb-9b86-61296cca34f0.jpg)

**Comment by ReinerBforartists** [comment]
Clean up panel

![cleanup](https://user-images.githubusercontent.com/13919702/119799121-13d8f400-bedc-11eb-8c36-db352d803210.jpg)

**Comment by ReinerBforartists** [comment]
Hm, the transform panel makes trouble with the several different modes and object types. So leaving the gaps here alone for now.

![consistent](https://user-images.githubusercontent.com/13919702/119802032-7d5a0200-bede-11eb-8578-e505ef3042af.jpg)

**Comment by ReinerBforartists** [comment]
Curve panel and tab

![curvepanel](https://user-images.githubusercontent.com/13919702/120061415-6a217080-c05d-11eb-8b17-dff25c9c1945.png)

**Comment by ReinerBforartists** [comment]
Control points

![controlpoints](https://user-images.githubusercontent.com/13919702/120061673-97225300-c05e-11eb-838d-ec70c4b0ae2b.jpg)

**Comment by ReinerBforartists** [comment]
![surfacemenu](https://user-images.githubusercontent.com/13919702/120064693-d86e2f00-c06d-11eb-83aa-ac1957f0dc88.jpg)

**Comment by ReinerBforartists** [comment]
Control points panel also for surface object

![cpointssurface](https://user-images.githubusercontent.com/13919702/120065138-2126e780-c070-11eb-8566-09eef421c09d.jpg)

**Comment by ReinerBforartists** [comment]
Regarding bl_context ...

source\blender\blenkernel\intern\context.c, line 1230

static const char *data_mode_strings[] = {
    "mesh_edit",           "curve_edit",          "surface_edit",        "text_edit",
    "armature_edit",       "mball_edit",          "lattice_edit",        "posemode",
    "sculpt_mode",         "weightpaint",         "vertexpaint",         "imagepaint",
    "particlemode",        "objectmode",          "greasepencil_paint",  "greasepencil_edit",
    "greasepencil_sculpt", "greasepencil_weight", "greasepencil_vertex", NULL,
};

**Comment by ReinerBforartists** [comment]
Mask tab in sculpt mode

![masktab](https://user-images.githubusercontent.com/13919702/120772835-03023100-c521-11eb-9a2a-c297f96c9dc0.jpg)

**Comment by ReinerBforartists** [comment]
Renamed the task. We don't add the BFA1 panels. Too different nowadays. But go directly from the existing menus.

**Comment by ReinerBforartists** [comment]
Initialize face sets panel

![init](https://user-images.githubusercontent.com/13919702/121144652-7caa5f80-c83e-11eb-8f70-680929831660.jpg)

**Comment by ReinerBforartists** [comment]
Vertex paint, paint panel

![paint](https://user-images.githubusercontent.com/13919702/121148241-e415de80-c841-11eb-8ea8-143b7e2974f1.jpg)

**Comment by ReinerBforartists** [comment]
Weight paint

![weights](https://user-images.githubusercontent.com/13919702/121159503-89818000-c84b-11eb-9a82-5a9b58b36b74.jpg)

**Comment by ReinerBforartists** [comment]
And that was it. I have added what made sense to me. There might be the one or another panel that could be in the tool shelf too. But i haven't added them for a reason. 

Let's work for a while with that system. And then we might decide for further changes if needed.

This task is done.
